### PR TITLE
Make n_step_lstm work with ndarray

### PR DESCRIPTION
**Pull request by unnonouno**
related to #1687

### CODE REVIEW

**Comment by okuta** [comment]
How about adding test case?

**Comment by unnonouno** [comment]
Currently all functions are not tested with ndarrays. We need to test such cases for all functions systematically.

**Comment by delta2323** [comment]
> We need to test such cases for all functions systematically.

FYI, we prepare the utilities that tests unary Chainer functions in #1534, although it does not cover n_step_lstm.

File: chainer/functions/connection/n_step_lstm.py [hunk 1]
```diff
@@ -453,7 +453,7 @@ def n_step_lstm(
 
     """
 
-    xp = cuda.get_array_module(hx.data)
+    xp = cuda.get_array_module(hx, hx.data)
```
Comment (delta2323) [inline chainer/functions/connection/n_step_lstm.py]: This change seems odd to me. Is it a correct one?

File: chainer/functions/connection/n_step_lstm.py [hunk 2]
```diff
@@ -483,13 +483,13 @@ def n_step_lstm(
...
         for x in xs:
-            batch = len(x.data)
+            batch = x.shape[0]
             h_next = []
             c_next = []
             for layer in six.moves.range(n_layers):
                 h = hx[layer]
                 c = cx[layer]
-                if len(h.data) > batch:
+                if h.shape[0] > batch:
```
Comment (delta2323) [inline chainer/functions/connection/n_step_lstm.py]: `len(h.shape)` is still valid. Is there any reason why change from `len()` to indexing?

File: chainer/functions/connection/n_step_lstm.py (hunk 1 above)
Comment (unnonouno) [inline chainer/functions/connection/n_step_lstm.py]: When `hx` is `Variable`, `hx.data` is ndarray.

Comment (unnonouno) [inline chainer/functions/connection/n_step_lstm.py]: I feel it is better to extend `get_rray_module` to support `Variable`

File: chainer/functions/connection/n_step_lstm.py (hunk 2 above)
Comment (unnonouno) [inline chainer/functions/connection/n_step_lstm.py]: Semantics of `len` of `Variable` is different from `ndarray`.

**Comment by delta2323** [comment]
Please fix the document of `NStepLSTM.__call__`.

File: chainer/functions/connection/n_step_lstm.py (hunk 1 above)
Comment (delta2323) [inline chainer/functions/connection/n_step_lstm.py]: I think it is good that `get_array_module` check the underlying arrays of `Variable`s. Could you make a PR?

Comment (delta2323) [inline chainer/functions/connection/n_step_lstm.py]: I did not understand that your intension is that it works `hx` is both `ndarray` or `Variable`. Now I come to think this change is acceptable.

File: chainer/functions/connection/n_step_lstm.py (hunk 2 above)
Comment (delta2323) [inline chainer/functions/connection/n_step_lstm.py]: OK, I understand it.

**Comment by delta2323** [comment]
I discussed the documentation with @unnonouno. As we do not test if `n_step_lstm` work with `ndarray`, we should not document the feature for now. So, we merge this feature without document and will add it after we introduce the systematic test mechanism of this feature.

**Pull request merged**